### PR TITLE
techdebt: [CDS-38672]: Adding reference for runtime user-groups

### PR DESCRIPTION
**Pull request by Y0709**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/34457)

### CODE REVIEW

**Comment by Y0709** [comment]
trigger messagecheck

**Comment by Y0709** [comment]
trigger messagecheck

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34457)* status: :shipit: complete! ✓

**Review by abhinav-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34457)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/PipelineServiceImpl.java` line 2107 at r1](https://reviewable.io/reviews/harness/harness-core/34457#-N5-6tEzDqi--2qFcZsP:-N5-6tEzDqi--2qFcZsQ:b-bdyay2) ([raw file](https://github.com/harness/harness-core/blob/9809d8e5d9abcba4cc3e96dda5ffcc402ee328db/400-rest/src/main/java/software/wings/service/impl/PipelineServiceImpl.java#L2107)):*
> ```Java
>         } else {
>           if (stageElement.getRuntimeInputsConfig() != null) {
>             List<String> userGroupIds = stageElement.getRuntimeInputsConfig().getUserGroupIds();
> ```

unit test?

File: 400-rest/src/main/java/software/wings/service/impl/PipelineServiceImpl.java [hunk 1]
```diff
@@ -2094,36 +2094,43 @@ public Set<String> getUserGroups(Pipeline pipeline) {
...
+          Map<String, Object> properties = stageElement.getProperties();
+          properties.forEach((name, value) -> {
+            if (USER_GROUPS.equals(name) && value instanceof List && isNotEmpty((List<String>) value)) {
+              userGroups.addAll((List<String>) value);
+            }
+          });
+        } else {
+          if (stageElement.getRuntimeInputsConfig() != null) {
+            List<String> userGroupIds = stageElement.getRuntimeInputsConfig().getUserGroupIds();
+            for (String id : userGroupIds) {
```
Comment (puthrayaharness) [inline 400-rest/src/main/java/software/wings/service/impl/PipelineServiceImpl.java:2108]: You can simply do a `userGroups.addAll(userGroupIds)`

File: 400-rest/src/main/java/software/wings/service/impl/PipelineServiceImpl.java [hunk 2]
```diff
@@ -2094,36 +2094,43 @@ public Set<String> getUserGroups(Pipeline pipeline) {
     }
     Set<String> userGroups = new HashSet<>();
     for (PipelineStage bean : pipeline.getPipelineStages()) {
-      PipelineStageElement stageElement = bean.getPipelineStageElements().get(0);
-      if (StateType.APPROVAL.name().equals(stageElement.getType())) {
-        Map<String, Object> properties = stageElement.getProperties();
-        properties.forEach((name, value) -> {
-          if (USER_GROUPS.equals(name) && value instanceof List && isNotEmpty((List<String>) value)) {
-            userGroups.addAll((List<String>) value);
+      for (PipelineStageElement stageElement : bean.getPipelineStageElements()) {
```
Comment (puthrayaharness) [inline 400-rest/src/main/java/software/wings/service/impl/PipelineServiceImpl.java:2097]: There is no need to change this logic? I don't think we have more than one pipelineStageElement.

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

**Comment by Y0709** [comment]
trigger messagecheck

**Review by puthrayaharness — APPROVED** [review]

**Review by puthrayaharness — APPROVED** [review]
> Reviewed 3 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34457)* status: :shipit: complete! ✓

**Pull request merged**